core/src/main/scala/zio/config/ReadModule.scala

```diff
@@ -268,19 +310,38 @@ private[config] trait ReadModule extends ConfigDescriptorModule {
     s"Provided value is ${given.toString}, expecting the type ${expectedType}"
 
   final def requiredZipAndOrFields[A](config: ConfigDescriptor[A]): Int = {
+    def countZipSize[A](config: ConfigDescriptor[A]): Int =
+      config match {
+        case ConfigDescriptorAdt.Zip(left, right) =>
+          countZipSize(left.get()) + countZipSize(right.get())
+        case _ => 1
+      }
```

Comment:
It was not straightforward how to change the logic in `handleDefaultValues`, but then I came up with this implementation. I'll try to reason about it.

The change is required because now `nested` terminates with an error if the subtree does not exists, while before it was just moving the context inner, and then each `Source` produced a separate error.

This function is only used in the following condition:
```
val baseConditionForFallBack =
      hasOnlyMissingValuesAndZeroIrrecoverable && sizeOfZipAndOrErrors(error) == requiredZipAndOrFields(config)
```
which basically tests that "everything is missing", in which case we replace the whole thing with a default value.

I changed the `Zip`, `OrElse` and `OrElseEither` cases. The idea is that each element of these is either:
- a `Source` in which case it counts as `1`
- a `Nested` in which case if the whole thing is missing, it terminates, so it counts as `1`. If it was only partially missing, it would not be `1` but then we have a count mismatch and the fallback condition is false which is good.
- Something else like `XmapEither`, or `Describe`, etc that does not affect the count

Because of this, I think that these "count a flat size" implementations provide the expected result for the above fallback condition (but only usable for that).

I can rewrite it in a more generic form as you suggested.

Reply:
👍